physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,721 @@
+		d.l.RUnlock()
+		return nil
+	}
+
+	servers := future.Configuration().Servers
+	serverIDs := make([]string, 0, len(servers))
+	for _, server := range servers {
+		serverIDs = append(serverIDs, string(server.ID))
+	}
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	ret := make(map[raft.ServerID]*autopilot.Server)
+	for id, state := range d.followerStates.followers {
+		// If the server is not in raft configuration, even if we received a follower
+		// heartbeat, it shouldn't be a known server for autopilot.
+		if !strutil.StrListContains(serverIDs, id) {
+			continue
+		}
+
+		server := &autopilot.Server{
+			ID:          raft.ServerID(id),
+			Name:        id,
+			RaftVersion: raft.ProtocolVersionMax,
+			Ext:         d.autopilotServerExt(state.Suffrage),
+		}
+
+		switch state.IsDead {
+		case true:
+			d.logger.Debug("informing autopilot that the node left", "id", id)
+			server.NodeStatus = autopilot.NodeLeft
+		default:
+			server.NodeStatus = autopilot.NodeAlive
+		}
+
+		ret[raft.ServerID(id)] = server
+	}
+
+	// Add the leader
+	ret[raft.ServerID(d.localID)] = &autopilot.Server{
+		ID:          raft.ServerID(d.localID),
+		Name:        d.localID,
+		RaftVersion: raft.ProtocolVersionMax,
+		NodeStatus:  autopilot.NodeAlive,
+		Ext:         d.autopilotServerExt("voter"),
+	}
+
+	return ret
+}
+
+// RemoveFailedServer is called by the autopilot library when it desires a node
+// to be removed from the raft configuration. This function removes the node
+// from the raft cluster and stops tracking its information in follower states.
+// This function needs to return quickly. Hence removal is performed in a
+// goroutine.
+func (d *Delegate) RemoveFailedServer(server *autopilot.Server) {
+	go func() {
+		d.logger.Info("removing dead server from raft configuration", "id", server.ID)
+		if future := d.raft.RemoveServer(server.ID, 0, 0); future.Error() != nil {
+			d.logger.Error("failed to remove server", "server_id", server.ID, "server_address", server.Address, "server_name", server.Name, "error", future.Error())
+		}
+		d.followerStates.Delete(string(server.ID))
+	}()
+}
+
+// SetFollowerStates sets the followerStates field in the backend to track peers
+// in the raft cluster.
+func (b *RaftBackend) SetFollowerStates(states *FollowerStates) {
+	b.l.Lock()
+	b.followerStates = states
+	b.l.Unlock()
+}
+
+// SetAutopilotConfig updates the autopilot configuration in the backend.
+func (b *RaftBackend) SetAutopilotConfig(config *AutopilotConfig) {
+	b.l.Lock()
+	b.autopilotConfig = config
+	b.logger.Info("updated autopilot configuration", "config", b.autopilotConfig)
+	b.l.Unlock()
+}
+
+// AutopilotConfig returns the autopilot configuration in the backend.
+func (b *RaftBackend) AutopilotConfig() *AutopilotConfig {
+	b.l.RLock()
+	conf := b.autopilotConfig
+	b.l.RUnlock()
+	return conf.Clone()
```

Comment:
Is it safe to call Clone without the lock?

Reply:
Yep. We were anyways returning the same values. We are now just returning it as a different reference. The only main goal here was that the caller doesn't inadvertently modify the config in the backend.

Reply:
It would be a tiny bit safer if you deferred the unlock instead, that way if someday code starts modifying the map in place, there won't be a race.